compiler/rustc_ast_passes/src/ast_validation.rs

```diff
@@ -213,6 +213,23 @@ impl<'a> AstValidator<'a> {
         err.emit();
     }
 
+    fn deny_anonymous_field(&self, field: &FieldDef, parent_span: Span, where_: &str) {
```

Comment:
This function should check for `field.ident.is_none()` instead of anonymous type.
Anonymous types need a separate check that would prohibit them everywhere except for anonymous fields.

Test cases:
```rust
fn f() -> struct { field: u8 } {} // Should error
struct S { field: struct { field: u8 } } // Should error
struct S { _: u8 } // Should error
```

Reply:
UPD: `struct S { _: SomeOtherStruct }` is actually legal (https://github.com/rust-lang/rfcs/blob/master/text/2102-unnamed-fields.md#unnamed-fields-with-named-types).

Reply:
I don't think I can check for `field.ident.is_none()`, because a `ItemKind::Struct` of `VariantData::Tuple` has fields with `None` as an identifier. Maybe I could pass directly a `kw::Underscore` as an identifier to differentiate between the two cases.

Reply:
Yes, keeping `Some(kw::Underscore)` should be better, that's what other users of `parse_ident_or_underscore` do.

Reply:
Just in case, I added a new test to check the rejection of `Unnamed(Struct, Union)` in various declarations. Maybe I missed some cases, but we should be able to add them to the test to ensure we cover all of them.